### PR TITLE
implement global variables + callbackUrl support

### PR DESCRIPTION
**Pull request by rubenmarcus**
# changes

- adds jest global mock file

**@mintbase-js/auth**
- remove unnecessary consts and add them to sdk package
- fix unit tests
- create a file to import the wallets we support
- create a file for the error messages

**@mintbase-js/data**
- added support to pass the network on all methods
- graphql fetch only happens if network is set
- removed unused consts

**@mintbase-js/sdk**

- implement `mbjs` namespace on sdk package with the `config` method and `keys` prop
- standardize arguments: 
   - `contractAddress` to all places with _nftContractId_, _contractId_ and _contractAddress_
   - _marketId_ now is `marketAddress`
 - remove repeated **consts**
 - add in order of priority:
   - method argument as env const
   - process.env | mbjs.keys
  - add support for process.env on node.js projects, and mbjs.config for client side 
  - some of the consts are enums   
 - fix unit tests
 - fix README.md files
 
 
 **@mintbase-js/rpc**
 - replaced all constants with mbjs.keys
 
 # tests
 
 **@mintbase-js/auth**
 - account and wallet methods ( modal and connect ) , should still work as it is, as also setting the vars: 
 NEAR_NETWORK on process.env, mbjs.config({ network: ...}) , connect(... network ) , will also set the current rpc url, ( can check with console.log(mbjs.keys) 

| method                    | .env | mbjs.config | method arg |
|---------------------------|------|-------------|------------|
| wallet                    |      |             |            |
| account                   |      |             |            |
 
 
 **@mintbase-js/data**
 all methods should work setting up the vars on process.env, on mbjs.config and on each test.
 
 | method                     | .env | mbjs.config | method arg |
|----------------------------|------|-------------|------------|
| attributesByMetaId         |      |             |            |
| ownedNftsByStore           |      |             |            |
| ownedStores                |      |             |            |
| ownedTokens                |      |             |            |
| storeData                  |      |             |            |
| tokenOwner                 |      |             |            |
| tokenById                  |      |             |            |
| tokenListingCountsByMetaId |      |             |            |
| tokenOwnersByMetadataId    |      |             |            |
| tokenProvenance            |      |             |            |
| tokensByStatus             |      |             |            |
 
 
 
 **@mintbase-js/sdk**
 
 | method                    | .env | mbjs.config | method arg |
|---------------------------|------|-------------|------------|
| addMinter                 |      |             |            |
| batchChangeMinters        |      |             |            |
| burn                      |      |             |            |
| buy                       |      |             |            |
| delist                    |      |             |            |
| deployContract            |      |             |            |
| depositStorage            |      |             |            |
| list                      |      |             |            |
| mint                      |      |             |            |
| removeMinter              |      |             |            |
| transfer                  |      |             |            |
| transferContractOwnership |      |             |            |
| v1/market.ts methods      |      |             |            |
| v1/token.ts methods       |      |             |            |

**@mintbase-js/react**

| method                    | .env | mbjs.config | method arg |
|---------------------------|------|-------------|------------|
| useOwnedNftsByStore       |      |             |            |
| useTokenById              |      |             |            |

**@mintbase-js/rpc**

| method                    | .env | mbjs.config | method arg |
|---------------------------|------|-------------|------------|
| getBalance                |      |             |            |
| getTxnStatus              |      |             |            |
| getBlockHeight            |      |             |            |

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203637195678052

### CODE REVIEW

File: packages/sdk/src/execute.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ export type NearContractCall = ContractCall | ContractCall[]
 export type NearCallSigningOptions = {
   wallet?: Wallet;
   account?: Account;
+  callbackUrl?: string;
```
Comment (cif) [inline packages/sdk/src/execute.ts:23]: NearCallSigningOptions are for methods used to sign the transaction, a callback URL should be passed as part of the NearContractCall

File: packages/sdk/src/execute.ts [hunk 2]
```diff
@@ -36,22 +38,26 @@ const validateSigningOptions = ({ wallet, account }: NearCallSigningOptions): vo
...
  */
 export const execute = async (
-  { wallet, account }: NearCallSigningOptions,
+  { wallet, account, callbackUrl }: NearCallSigningOptions,
   ...calls: NearContractCall[]
 ): Promise<void | providers.FinalExecutionOutcome | providers.FinalExecutionOutcome[] > => {
   validateSigningOptions({ wallet, account });
 
-  const outcomes = await genericBatchExecute(flattenArgs(calls), wallet, account);
+  console.log(callbackUrl, 'callbackUrl');
```
Comment (cif) [inline packages/sdk/src/execute.ts:46]: 🧹

File: packages/sdk/src/execute.ts [hunk 3]
```diff
@@ -36,22 +38,26 @@ const validateSigningOptions = ({ wallet, account }: NearCallSigningOptions): vo
...
   }
   return outcomes;
 
 };
 
-const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account: Account): Promise<void | providers.FinalExecutionOutcome[]> =>{
+const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account: Account, callbackUrl: string): Promise<void | providers.FinalExecutionOutcome[]> =>{
   if (wallet) {
-    return batchExecuteWithBrowserWallet(call, wallet);
+    return batchExecuteWithBrowserWallet(call, wallet, callbackUrl);
```
[the remark this answers is not in the document]
Comment (cif) [reply packages/sdk/src/execute.ts:60]: Or think of about another way to pass it without using the first argument, but nothing coming quickly to mind for me

[the remark this answers is not in the document]
Comment (rubenmarcus) [reply packages/sdk/src/execute.ts:64]: @cif  `window.mbJsConfig.callbackUrl ` 🤣 🤣 🤣 🤣

File: packages/data/src/constants.ts [hunk 4]
```diff
@@ -16,8 +18,14 @@ if (globalThis?.window?.location.host.includes(DataNetwork.SANDBOX)) {
   hostBasedNetworkDefault = DataNetwork.SANDBOX;
 }
 
-export const NEAR_DATA_ENV = process.env.NEAR_DATA_ENV ||
-  hostBasedNetworkDefault;
 
-export const GRAPHQL_ENDPOINT = `https://interop-${NEAR_DATA_ENV}.hasura.app/v1/graphql`;
+// export const NEAR_DATA_ENV = mbjs.envs.network ||  hostBasedNetworkDefault;
+
+// console.log(mbjs, mbjs.envs, mbjs.envs.network, NEAR_DATA_ENV,  'data package mbjs import test');
```
Comment (sainthiago) [inline packages/data/src/constants.ts:24]: 🧹

File: packages/data/src/constants.ts [hunk 5]
```diff
@@ -16,8 +18,14 @@ if (globalThis?.window?.location.host.includes(DataNetwork.SANDBOX)) {
...
+
+
+export const GRAPHQL_ENDPOINT = 'https://interop-rubennet.hasura.app/v1/graphql';
+
+// console.log(GRAPHQL_ENDPOINT, 'graphql endpoint');
```
Comment (sainthiago) [inline packages/data/src/constants.ts:29]: 🧹

File: packages/data/src/graphql/fetch.ts [hunk 6]
```diff
@@ -12,19 +16,38 @@ export type GqlFetchResult<T> = {
...
-    };
-
-  } catch (error:unknown) {
-    return {
-      error: getErrorMessage(error),
-    };
+
+  const endpointReady =  window.mintbase.isSet || env?.length > 0; 
+
+  console.log(endpointReady, mbjs?.keys, window.mintbase , 'endpointReady');
```
Comment (sainthiago) [inline packages/data/src/graphql/fetch.ts:28]: 🧹

File: packages/sdk/src/config.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      isSet: true,
+    }; 
+
+    CONFIG_KEYS.network = globalConfig.network;
+    CONFIG_KEYS.graphql_url = globalConfig.graphqlUrl;
+    CONFIG_KEYS.callbackUrl = globalConfig.callbackUrl;
+    CONFIG_KEYS.isSet = globalConfig.isSet;
+
+
+    console.log(globalConfig, CONFIG_KEYS, 'keys');
```
Comment (sainthiago) [inline packages/sdk/src/config.ts:33]: 🧹

File: packages/sdk/src/execute.ts [hunk 8]
```diff
@@ -31,29 +33,36 @@ const validateSigningOptions = ({ wallet, account }: NearCallSigningOptions): vo
...
  */
 export const execute = async (
-  { wallet, account }: NearCallSigningOptions,
+  { wallet, account,   callbackUrl }: NearExecuteOptions,
   ...calls: NearContractCall[]
 ): Promise<void | providers.FinalExecutionOutcome | providers.FinalExecutionOutcome[] > => {
   validateSigningOptions({ wallet, account });
 
-  const outcomes = await genericBatchExecute(flattenArgs(calls), wallet, account);
+  console.log(callbackUrl, 'callbackUrl');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:46]: 🧹

File: packages/sdk/src/execute.ts [hunk 9]
```diff
@@ -31,29 +33,36 @@ const validateSigningOptions = ({ wallet, account }: NearCallSigningOptions): vo
...
+
+  console.log('wallet', wallet);
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:48]: 🧹

File: packages/sdk/src/execute.ts [hunk 10]
```diff
@@ -31,29 +33,36 @@ const validateSigningOptions = ({ wallet, account }: NearCallSigningOptions): vo
...
     return outcomes[0];
   }
   return outcomes;
 
 };
 
-const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account: Account): Promise<void | providers.FinalExecutionOutcome[]> =>{
+const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account: Account, callbackUrl: string): Promise<void | providers.FinalExecutionOutcome[]> =>{
+
+  console.log(callbackUrl, 'callbackUrl genericBatchExecute');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:60]: 🧹

File: packages/sdk/src/execute.ts [hunk 11]
```diff
@@ -63,16 +72,21 @@ const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account
 const batchExecuteWithNearAccount = async (
   calls: ContractCall[],
   account: Account,
+  callbackUrl: string,
 ): Promise<FinalExecutionOutcome[]> => {
-  const outcomes: FinalExecutionOutcome[] =[];
+  const outcomes: any[] =[];
   for (const call of calls) {
+    console.log(call , 'batchExecuteWithNearAccountbatchExecuteWithNearAccount call');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:79]: 🧹

File: packages/sdk/src/execute.ts [hunk 12]
```diff
@@ -63,16 +72,21 @@ const genericBatchExecute = async (call: ContractCall[], wallet: Wallet, account
...
+    console.log(callbackUrl , 'batchExecuteWithNearAccount callbackUrl');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:80]: 🧹

File: packages/sdk/src/execute.ts [hunk 13]
```diff
@@ -86,9 +100,16 @@ const batchExecuteWithNearAccount = async (
 const batchExecuteWithBrowserWallet = async (
   calls: ContractCall[],
   wallet: Wallet,
+  callback: string,
 ): Promise<void | FinalExecutionOutcome[]> => {
+
+  console.log(wallet, 'wallet  batchExecuteWithBrowserWallet');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:106]: 🧹

File: packages/sdk/src/execute.ts [hunk 14]
```diff
@@ -86,9 +100,16 @@ const batchExecuteWithNearAccount = async (
...
+  console.log(callback, 'callback  batchExecuteWithBrowserWallet');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:107]: 🧹

File: packages/sdk/src/execute.ts [hunk 15]
```diff
@@ -98,20 +119,26 @@ declare type TxnOptionalSignerId = Optional<Transaction, 'signerId'>;
...
-      methodName: call.methodName,
-      args: call.args,
-      gas: call.gas as string,
-      deposit: call.deposit as string,
-    },
-  }],
-});
+): BrowserWalletSignAndSendTransactionParams | TxnOptionalSignerId => {
+
+  console.log(call , 'convertGenericCallToWalletCall call');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:124]: 🧹

File: packages/sdk/src/execute.ts [hunk 16]
```diff
@@ -98,20 +119,26 @@ declare type TxnOptionalSignerId = Optional<Transaction, 'signerId'>;
...
+  console.log(call.callbackUrl , 'convertGenericCallToWalletCall callbackUrl');
```
Comment (sainthiago) [inline packages/sdk/src/execute.ts:125]: 🧹

**Review by sainthiago — CHANGES_REQUESTED** [review]

File: packages/app/README.md [hunk 17]
```diff
@@ -2,4 +2,43 @@ This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next
 
 ## Getting Started
 
-Mintbase JS Browser/Next testing suite. 
\ No newline at end of file
+Mintbase JS Browser/Next testing suite. 
+
+# config vars
+
+You can set and NEED to set the Near Network you gonna use on your app in three ways:
```
Comment (cif) [inline packages/app/README.md:9]:
```suggestion
You may specify the NEAR network you want to interact with in three ways:
```

File: packages/auth/README.md [hunk 18]
```diff
@@ -15,6 +15,46 @@ This module can be used as a framework agnostic wrapper using framework specific
 
 *Check back soon for documentation on wallet selector wrapper methods and example usage with other frameworks in the near future.*
 
+
+# config vars
+
+You can set and NEED to set the Near Network you gonna use on your app in three ways:
```
Comment (cif) [inline packages/auth/README.md:21]: Same as above suggestion, do we need to specify this in both objects or can we do this in a single core lib then access from the others?

File: packages/auth/src/wallet.ts [hunk 19]
```diff
@@ -50,27 +36,17 @@ export let walletSelectorComponents: WalletSelectorComponents  = {
...
     ],
   });
 
-  const defaultMintbaseContract = NEAR_NETWORK === 'testnet'
-    ? DEFAULT_MINTBASE_CONTRACT_TESTNET
-    : DEFAULT_MINTBASE_CONTRACT_MAINNET;
 
   const modal = setupModal(selector, {
-    contractId: NEAR_LOGIN_CONTRACT_ID || defaultMintbaseContract,
+    contractId: mbjs.keys.contractAddress ?? mbjs.keys.mbContract,
```
Comment (cif) [inline packages/auth/src/wallet.ts:49]: To clarify your question asked in dev channel regarding this option, when you login with a contract you are giving an allowance of 0.25 N that can be used for calls on that contract without signing.  We enforce signing for nearly all transactions anyway, so it's not the most relevant to set - but the user does see that when logging in so it's nice to have it configurable.

File: packages/data/src/graphql/fetch.ts [hunk 20]
```diff
@@ -12,19 +12,36 @@ export type GqlFetchResult<T> = {
...
-  } catch (error:unknown) {
-    return {
-      error: getErrorMessage(error),
-    };
+
+  const endpointReady =  mbjs?.keys?.isSet || network?.length > 0; 
+
+  let graphqlEndpoint = mbjs?.keys.graphqlUrl ?? '';
+
+  if (network && network.length > 0) {
```
Comment (cif) [inline packages/data/src/graphql/fetch.ts:26]: What's the motivation for empty string here vs null or undef?

File: packages/data/src/graphql/fetch.ts [hunk 21]
```diff
@@ -12,19 +12,36 @@ export type GqlFetchResult<T> = {
...
+    graphqlEndpoint = GRAPHQL_ENDPOINTS[network];
+  }
+
+  if (endpointReady) {
```
Comment (cif) [inline packages/data/src/graphql/fetch.ts:30]: Overall the logic in here is a bit hard to follow.  If we don't have an endpoint configured correctly, we only need one logical operation `if(!graphQLEndpoint) { throw new Error('GraphQL endpoint not configured...'); }`. this should be unit tested as well

File: packages/data/src/api/tokensByStatus/tokensByStatus.ts [hunk 22]
```diff
@@ -1,17 +1,19 @@
+import { Network } from '@mintbase-js/sdk';
 import { fetchGraphQl } from '../../graphql/fetch';
 import { ParsedDataReturn } from '../../types';
 import { parseData, getTokenArrayFromNodes } from '../../utils';
 import { TokenByStatusQueryResults, TokensByStatus } from './tokenByStatus.types';
 import { tokensByStatusQuery } from './tokensByStatus.query';
 
 
-export const tokensByStatus = async (metadataId: string, ownedBy?: string): Promise<ParsedDataReturn<TokensByStatus>> => {
+export const tokensByStatus = async (metadataId: string, ownedBy?: string,network?: Network ): Promise<ParsedDataReturn<TokensByStatus>> => {
```
Comment (cif) [inline packages/data/src/api/tokensByStatus/tokensByStatus.ts:9]: Nit picky, but space between args

File: packages/react/src/hooks/useTokenById.ts [hunk 23]
```diff
@@ -16,12 +17,17 @@ export const useTokenById = (
...
+
+  const contract = contractAddress || mbjs.keys.contractAddress;
+  const validParams = contract && tokenId; 
+
+
   useEffect(() => {
     let isCancelled = false;
 
-    if (loading) {
+    if (loading && validParams) {
```
Comment (cif) [inline packages/react/src/hooks/useTokenById.ts:28]: Even in the context of a hook/effect, an early return is always easier to read:

```
if (!something) {
  setError('something needs to be defined')
  return
}
```

vs wrapping later and then noop

**Review by cif — APPROVED** [review]
Overall good enough to merge, would prefer null or undef vs empty string perhaps this is some side effect of how the config is being loaded?

**Review by sainthiago — APPROVED** [review]

File: packages/data/src/graphql/fetch.ts (hunk 21 above)
Comment (rubenmarcus) [reply packages/data/src/graphql/fetch.ts]: so[ changed here](https://github.com/Mintbase/mintbase-js/blob/a6f94130ae6954ef6e6356f497403392c1810b16/packages/data/src/graphql/fetch.ts)


added the logic to check:

- is valid network coming from argument or mbjs.keys.network
  - if not ,return error message
- if network isnt set in any of the places
  - if not return error message
- if endpointReady = network exists and is valid, it will set graphqEndpoint
- if endpointReady && graphqlEndpoint === fetch the data

File: packages/react/src/hooks/useTokenById.ts (hunk 23 above)
Comment (rubenmarcus) [reply packages/react/src/hooks/useTokenById.ts:28]: @sainthiago 

`useEffect expects the return to be a clean up function, so return null is wrong because 1. return is conditional and 2. return is not a function`

this type of code we being doing on omni-site is wrong

more info:
[What is the expected return of `useEffect` used for?](https://stackoverflow.com/questions/56800694/what-is-the-expected-return-of-useeffect-used-for)

[Cleanup useEffect](https://hackernoon.com/cleanup-functions-in-reacts-useeffect-hook-explained)
[Understanding React’s useEffect cleanup function](https://blog.logrocket.com/understanding-react-useeffect-cleanup-function/)

File: packages/data/src/graphql/fetch.ts (hunk 20 above)
Comment (rubenmarcus) [reply packages/data/src/graphql/fetch.ts]: fixed

File: packages/react/src/hooks/useTokenById.ts (hunk 23 above)
Comment (cif) [reply packages/react/src/hooks/useTokenById.ts:28]: We could define the return function in the outer scope if we cared that much, but I think this is fine given it's a hook

**Comment by rubenmarcus** [comment]
tested on local react-vite project:

pass on 
`await tokenById( '5','logout.mintbase1.near', 'mainnet');`. ✅ 
`mbjs.config({network: 'mainnet', callbackUrl: 'mintbase.xyz'}) `✅ 


tested on local next.js 13 project

`await tokenById( '5','logout.mintbase1.near', 'mainnet');`. ✅ 
`mbjs.config({network: 'mainnet', callbackUrl: 'mintbase.xyz'}) `✅ 

`process.env.NEAR_NETWORK`. only on server side ✅

File: packages/auth/src/wallet.ts (hunk 19 above)
Comment (rubenmarcus) [reply packages/auth/src/wallet.ts:52]: followed the previous logic
